scripts/create-darwin-volume.sh

```diff
@@ -127,42 +164,75 @@ main() {
-                exit 1
-            fi
+    if [ -n "$create_volume" ]; then
+        echo "Creating a Nix volume..." >&2
+
+        sudo diskutil apfs addVolume "$disk" "$NIX_VOLUME_FS" "$volume" -mountpoint /nix
+        new_uuid="$(volume_uuid "$volume")"
+
+        if [ "$INSTALL_MODE" = "no-daemon" ]; then # exported by caller
+            # TODO: is there a better way to do this?
+            sudo chown $USER:admin /nix
         fi
 
-        sudo diskutil apfs addVolume "$disk" APFS 'Nix Store' -mountpoint /nix
-        volume="Nix Store"
+        if test_filevault_in_use; then
+            # security program's flags won't let us both specify a keychain
+            # and be prompted for a pw to add; two step workaround:
+            # 1. add a blank pw to system keychain
+
+            # system is in some sense less secure than user keychain... (it's
+            # possible to read the password for decrypting the keychain) but
+            # the user keychain appears to be available too late. As far as I
+            # can tell, the file with this password (/var/db/SystemKey) is
+            # inside the FileVault envelope. If that isn't true, it may make
+            # sense to store the password inside the envelope?
+            sudo /usr/bin/security add-generic-password -a "$volume" -s "$new_uuid" -D "$volume encryption password" -j "Added automatically by the Nix installer for use by /Library/LaunchDaemons/org.nixos.darwin-store.plist" "/Library/Keychains/System.keychain"
+            # TODO: decide if we should add `-T /System/Library/CoreServices/APFSUserAgent`
+            # This should let the system seamlessly supply the password for this volume
+            # which in turn means the fstab entry is enough for the system to (eventually)
+            # decrypt and mount the volume we're adding, but I hesitate because I'm not
+            # certain the system _should_ transparently failover if the LaunchDaemon is
+            # broken for some reason? Without supplying this flag, the system will instead
+            # start prompting them to allow APFSUserAgent to access this credential.
+
+            # 2. add a password with the -U (update) flag and -w (prompt if last)
+            #    flags, but specify no keychain; security will use the first it finds
+            prepare_darwin_volume_password "$volume" "$new_uuid" | sudo diskutil apfs encryptVolume "$volume" -user disk -stdinpassphrase
+        fi
     else
         echo "Using existing '$volume' volume" >&2
     fi
 
-    if ! test_fstab; then
-        echo "Configuring /etc/fstab..." >&2
-        label=$(echo "$volume" | sed 's/ /\\040/g')
-        # shellcheck disable=SC2209
-        printf "\$a\nLABEL=%s /nix apfs rw,nobrowse\n.\nwq\n" "$label" | EDITOR=ed sudo vifs
+    if ! test_voldaemon; then
+        echo "Configuring LaunchDaemon to mount '$volume'..." >&2
+        generate_mount_daemon | sudo tee /Library/LaunchDaemons/org.nixos.darwin-store.plist >/dev/null
```

Comment:
As previously mentioned, we need to pipe the volume name here so it can write the correct plist.

This does raise the worry of "what if we detect the plist already exists, but it has the wrong volume name" though.

Reply:
I'm inclined to say if we detect that the plist already exists, we should unload it (`sudo launchctl bootout system/org.nixos.darwin-store`, note that this may return a non-zero exit code if the process is currently running but it will boot it out anyway according to my own experience), delete it, and then recreate it. This way we can be sure to embed the correct info.

Or at least, we should do this if `create_volume` is set. If it's not set, then we're re-using an existing volume and presumably the existing plist (if there is one) is therefore still valid.

Reply:
This is something I left ~unhandled for discussion once I realized I'd need to thread it through everything (though I'm focused on the big-picture question about single-user mode before some of these narrower implementation questions).

My preference is to support a single fixed name (and maybe support pulling overriding it via the environment, if someone's really determined to do their own thing.) There are already a lot of moving parts here, and I actually think it's a kindness to tell the user "If you want to prepare your own volume, just name/label it 'Nix'"?

Reply:
If we pass the `-s` flag to `find-generic-password` (and I think we should) then that's not a fixed string, that's the volume UUID. I think this is important because, as previously stated, if I uninstall Nix but forget to delete the keychain entry, reinstalling it will give me two keychain entries with the same account name but different services, and we need to pick the right one.

Since we have to embed that UUID in the launch daemon, that means we do need to delete and recreate the plist rather than saying "oh it already exists, leave it alone".

In the case where the user prepares their own volume, we should just not touch the launch daemon at all. Either the user is re-reusing a previous volume (at which point the previous launch daemon applies), or they made their own custom volume and the launch daemon shouldn't exist at all. Or to put it another way, we should only touch the launch daemon when we add an entry to the keychain.